### PR TITLE
HPCC-12810 Include dmetaphone in the OSS version of HPCC Systems

### PR DESCRIPTION
**Pull request by richardkchapman**
Moved from the EE builds - this has been placed in the public domain by the
original author.

Signed-off-by: Richard Chapman rchapman@hpccsystems.com

### CODE REVIEW

**Comment by richardkchapman** [comment]
See related pr https://github.com/hpcc-systems/LN/pull/718

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: ecllibrary/teststd/Metaphone/TestMetaphone.ecl [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.  All rights reserved.
+############################################################################## */
+IMPORT Std.Metaphone;
+
+EXPORT TestMetaphone := MODULE
+
+  EXPORT TestConst := MODULE
+    EXPORT Test01 := ASSERT(Metaphone.primary('Algernon') = 'ALKRNN');
+    EXPORT Test01 := ASSERT(Metaphone.secondary('Algernon') = 'ALJRNN');
```
Comment (ghalliday) [inline ecllibrary/teststd/Metaphone/TestMetaphone.ecl]: This won't actually compile - the EXPORTS need to have different names.

**Comment by ghalliday** [comment]
This really should be using the StringBuffer class rather than a new cString class which does almost exactly the same.

File: plugins/dmetaphone/metaphone.cpp [hunk 2]
```diff
@@ -0,0 +1,893 @@
...
+    if (start < 0) return false;
+
+    char    target[64];
+    char*   test;
+
+    if (Len - start < len)
+    {
+        return false;
+    }
+    memcpy( target, Ptr + start, len );
```
Comment (ghalliday) [inline plugins/dmetaphone/metaphone.cpp]: this copy could probably be avoided.

File: plugins/dmetaphone/metaphone.cpp [hunk 3]
```diff
@@ -0,0 +1,893 @@
...
+////////////////////////////////////////////////////////////////////////////////
+bool MString::StringAt(int start, int len, ... )
+{
+
+    if (start < 0) return false;
+
+    char    target[64];
+    char*   test;
+
+    if (Len - start < len)
```
Comment (ghalliday) [inline plugins/dmetaphone/metaphone.cpp]: Variables called len and Len are a recipe for confusion!

**Comment by richardkchapman** [comment]
It's not a new cString class, it's a very old one (also used by some other legacy plugins).

We could port to use StringBuffer but you could also argue that is a separate issue from the "move from EE" issue this is fixing. Though there's certainly something to be said for fixing at the same time so cString does not need to be moved over.

File: plugins/dmetaphone/metaphone.cpp (hunk 3 above)
Comment (richardkchapman) [inline plugins/dmetaphone/metaphone.cpp]: Indeed. But this is not our code...

**Comment by ghalliday** [comment]
@richardkchapman only merge issue is the export.  

A separate question is do we treat the metaphone code as a black box, or do we want to treat it as our own code.  If the latter then we would need more test cases, should probably remove cString, and there is quite a lot of code that could be cleaned up.  Would be a different issue though.

**Comment by richardkchapman** [comment]
Given it's code that has not changed for 12 years (afaik) I'm inclined to leave it alone. It's hard to justify the work (and the risk) in terms of reduced future maintenance given how much maintenance this has had...

File: ecllibrary/std/Metaphone.ecl [hunk 4]
```diff
@@ -0,0 +1,41 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2015 HPCC Systems.  All rights reserved.
```
Comment (ghalliday) [inline ecllibrary/std/Metaphone.ecl]: I know this header matches the others in ecllibrary - but is it correct?

Comment (richardkchapman) [inline ecllibrary/std/Metaphone.ecl]: Why would it not be? This is not part of the original dmetaphone code.

File: plugins/dmetaphone/dmetaphone.cpp [hunk 5]
```diff
@@ -0,0 +1,120 @@
+#include "platform.h"
+#include <time.h>
+#include <stdlib.h>
+#include <string.h>
+#include "dmetaphone.hpp"
+#include "metaphone.h"
+
+static char buildVersion[] = "$HeadURL: https://svn.br.seisint.com/ecl/trunk/ln/plugins/dmetaphone/dmetaphone.cpp $ $Id: dmetaphone.cpp 60037 2010-10-07 13:00:46Z ghalliday $";
```
Comment (richardkchapman) [inline plugins/dmetaphone/dmetaphone.cpp]: Should probably be removed

File: ecllibrary/std/Metaphone.ecl (hunk 4 above)
Comment (richardkchapman) [inline ecllibrary/std/Metaphone.ecl]: If it needs more then that should be a Jira for the whole of ecllibrary

**Pull request merged**